Storage/Bookmarks/BookmarksModel.swift

```diff
@@ -62,6 +62,11 @@ public class BookmarkFolder: BookmarkNode {
     override public var canDelete: Bool {
         return false
     }
+
+    public func removeItemWithGUID(guid: GUID) -> MemoryBookmarkFolder {
```

Comment:
> Indeed, I wonder if the language constraint that forced us to make BookmarkFolder a class rather than a protocol still applies?

Ya 'abstract' classes seem like a smell in Swift. If we can flatten the hierarchy using protocols that seems more sensible and extendible. 

I gave it a shot over here: https://github.com/mozilla-mobile/firefox-ios/commit/6e9d871efb6a0f6764da954a9339c3dba9f7889e#diff-9d58e2ab1ac6a5e9c80ec12a3f087948

I think the language constraint you might have run into is not being able to add extensions to protocols for default implementations.

This looks like it works pretty well except there's this odd `BookmarkFolder` being created here: https://github.com/mozilla-mobile/firefox-ios/commit/6e9d871efb6a0f6764da954a9339c3dba9f7889e#diff-f01b23db37df0f5be33d74ce21acc51dR628

How come we're creating an instance of the 'abstract' class? Should this be a SQLiteBookmarkFolder?